File: model/errors_test.go
// This file is part of the happyDomain (R) project.
// Copyright (c) 2020-2025 happyDomain
// Authors: Pierre-Olivier Mercier, et al.
//
// This program is offered under a commercial and under the AGPL license.
// For commercial licensing, contact us at <contact@happydomain.org>.
//
// For AGPL licensing:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

package happydns_test

import (
	"errors"
	"net/http"
	"testing"

	"git.happydns.org/happyDomain/model"
)

func TestCustomErrorError(t *testing.T) {
	baseErr := errors.New("test error message")
	customErr := happydns.CustomError{
		Err:      baseErr,
		UserLink: "https://example.com/help",
		Status:   http.StatusBadRequest,
	}

	if customErr.Error() != "test error message" {
		t.Errorf("CustomError.Error() = %q; want %q", customErr.Error(), "test error message")
	}
}

func TestCustomErrorToErrorResponse(t *testing.T) {
	baseErr := errors.New("custom error")
	customErr := happydns.CustomError{
		Err:      baseErr,
		UserLink: "https://example.com/docs",
		Status:   http.StatusBadRequest,
	}

	resp := customErr.ToErrorResponse()

	if resp.Message != "custom error" {
		t.Errorf("ToErrorResponse().Message = %q; want %q", resp.Message, "custom error")
	}

	if resp.Link != "https://example.com/docs" {
		t.Errorf("ToErrorResponse().Link = %q; want %q", resp.Link, "https://example.com/docs")
	}
}

func TestCustomErrorHTTPStatus(t *testing.T) {
	tests := []struct {
		name           string
		status         int
		expectedStatus int
	}{
		{
			name:           "bad request",
			status:         http.StatusBadRequest,
			expectedStatus: http.StatusBadRequest,
		},
		{
			name:           "not found",
			status:         http.StatusNotFound,
			expectedStatus: http.StatusNotFound,
		},
		{
			name:           "internal server error",
			status:         http.StatusInternalServerError,
			expectedStatus: http.StatusInternalServerError,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			customErr := happydns.CustomError{
				Err:    errors.New("test"),
				Status: tt.status,
			}

			if customErr.HTTPStatus() != tt.expectedStatus {
				t.Errorf("HTTPStatus() = %d; want %d", customErr.HTTPStatus(), tt.expectedStatus)
			}
		})
	}
}

func TestForbiddenErrorError(t *testing.T) {
	forbiddenErr := happydns.ForbiddenError{
		Msg: "access denied",
	}

	if forbiddenErr.Error() != "access denied" {
		t.Errorf("ForbiddenError.Error() = %q; want %q", forbiddenErr.Error(), "access denied")
	}
}

func TestForbiddenErrorToErrorResponse(t *testing.T) {
	forbiddenErr := happydns.ForbiddenError{
		Msg: "forbidden resource",
	}

	resp := forbiddenErr.ToErrorResponse()

	if resp.Message != "forbidden resource" {
		t.Errorf("ToErrorResponse().Message = %q; want %q", resp.Message, "forbidden resource")
	}

	if resp.Link != "" {
		t.Errorf("ToErrorResponse().Link = %q; want empty string", resp.Link)
	}
}

func TestForbiddenErrorHTTPStatus(t *testing.T) {
	forbiddenErr := happydns.ForbiddenError{
		Msg: "test",
	}

	if forbiddenErr.HTTPStatus() != http.StatusForbidden {
		t.Errorf("HTTPStatus() = %d; want %d", forbiddenErr.HTTPStatus(), http.StatusForbidden)
	}
}

func TestInternalErrorError(t *testing.T) {
	baseErr := errors.New("internal error message")
	internalErr := happydns.InternalError{
		Err:         baseErr,
		UserMessage: "something went wrong",
		UserLink:    "https://example.com/status",
	}

	if internalErr.Error() != "internal error message" {
		t.Errorf("InternalError.Error() = %q; want %q", internalErr.Error(), "internal error message")
	}
}

func TestInternalErrorToErrorResponse(t *testing.T) {
	tests := []struct {
		name            string
		err             error
		userMessage     string
		userLink        string
		expectedMessage string
		expectedLink    string
	}{
		{
			name:            "with user message",
			err:             errors.New("internal error"),
			userMessage:     "user-friendly message",
			userLink:        "https://example.com/help",
			expectedMessage: "user-friendly message",
			expectedLink:    "https://example.com/help",
		},
		{
			name:            "without user message",
			err:             errors.New("internal error"),
			userMessage:     "",
			userLink:        "https://example.com/help",
			expectedMessage: "internal error",
			expectedLink:    "https://example.com/help",
		},
		{
			name:            "no link",
			err:             errors.New("error"),
			userMessage:     "user message",
			userLink:        "",
			expectedMessage: "user message",
			expectedLink:    "",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			internalErr := happydns.InternalError{
				Err:         tt.err,
				UserMessage: tt.userMessage,
				UserLink:    tt.userLink,
			}

			resp := internalErr.ToErrorResponse()

			if resp.Message != tt.expectedMessage {
				t.Errorf("ToErrorResponse().Message = %q; want %q", resp.Message, tt.expectedMessage)
			}

			if resp.Link != tt.expectedLink {
				t.Errorf("ToErrorResponse().Link = %q; want %q", resp.Link, tt.expectedLink)
			}
		})
	}
}

func TestInternalErrorHTTPStatus(t *testing.T) {
	internalErr := happydns.InternalError{
		Err: errors.New("test"),
	}

	if internalErr.HTTPStatus() != http.StatusInternalServerError {
		t.Errorf("HTTPStatus() = %d; want %d", internalErr.HTTPStatus(), http.StatusInternalServerError)
	}
}

func TestNotFoundErrorError(t *testing.T) {
	notFoundErr := happydns.NotFoundError{
		Msg: "resource not found",
	}

	if notFoundErr.Error() != "resource not found" {
		t.Errorf("NotFoundError.Error() = %q; want %q", notFoundErr.Error(), "resource not found")
	}
}

func TestNotFoundErrorToErrorResponse(t *testing.T) {
	notFoundErr := happydns.NotFoundError{
		Msg: "page not found",
	}

	resp := notFoundErr.ToErrorResponse()

	if resp.Message != "page not found" {
		t.Errorf("ToErrorResponse().Message = %q; want %q", resp.Message, "page not found")
	}

	if resp.Link != "" {
		t.Errorf("ToErrorResponse().Link = %q; want empty string", resp.Link)
	}
}

func TestNotFoundErrorHTTPStatus(t *testing.T) {
	notFoundErr := happydns.NotFoundError{
		Msg: "test",
	}

	if notFoundErr.HTTPStatus() != http.StatusNotFound {
		t.Errorf("HTTPStatus() = %d; want %d", notFoundErr.HTTPStatus(), http.StatusNotFound)
	}
}

func TestValidationErrorError(t *testing.T) {
	validationErr := happydns.ValidationError{
		Msg: "validation failed",
	}

	if validationErr.Error() != "validation failed" {
		t.Errorf("ValidationError.Error() = %q; want %q", validationErr.Error(), "validation failed")
	}
}

func TestValidationErrorToErrorResponse(t *testing.T) {
	validationErr := happydns.ValidationError{
		Msg: "invalid input",
	}

	resp := validationErr.ToErrorResponse()

	if resp.Message != "invalid input" {
		t.Errorf("ToErrorResponse().Message = %q; want %q", resp.Message, "invalid input")
	}

	if resp.Link != "" {
		t.Errorf("ToErrorResponse().Link = %q; want empty string", resp.Link)
	}
}

func TestValidationErrorHTTPStatus(t *testing.T) {
	validationErr := happydns.ValidationError{
		Msg: "test",
	}

	if validationErr.HTTPStatus() != http.StatusBadRequest {
		t.Errorf("HTTPStatus() = %d; want %d", validationErr.HTTPStatus(), http.StatusBadRequest)
	}
}

func TestHTTPErrorInterface(t *testing.T) {
	tests := []struct {
		name     string
		err      happydns.HTTPError
		wantCode int
	}{
		{
			name: "CustomError",
			err: happydns.CustomError{
				Err:    errors.New("test"),
				Status: http.StatusTeapot,
			},
			wantCode: http.StatusTeapot,
		},
		{
			name: "ForbiddenError",
			err: happydns.ForbiddenError{
				Msg: "forbidden",
			},
			wantCode: http.StatusForbidden,
		},
		{
			name: "InternalError",
			err: happydns.InternalError{
				Err: errors.New("test"),
			},
			wantCode: http.StatusInternalServerError,
		},
		{
			name: "NotFoundError",
			err: happydns.NotFoundError{
				Msg: "not found",
			},
			wantCode: http.StatusNotFound,
		},
		{
			name: "ValidationError",
			err: happydns.ValidationError{
				Msg: "invalid",
			},
			wantCode: http.StatusBadRequest,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.err.HTTPStatus() != tt.wantCode {
				t.Errorf("HTTPStatus() = %d; want %d", tt.err.HTTPStatus(), tt.wantCode)
			}

			resp := tt.err.ToErrorResponse()
			if resp.Message == "" {
				t.Error("ToErrorResponse().Message should not be empty")
			}
		})
	}
}

func TestPredefinedErrors(t *testing.T) {
	tests := []struct {
		name string
		err  error
		msg  string
	}{
		{
			name: "ErrAuthUserNotFound",
			err:  happydns.ErrAuthUserNotFound,
			msg:  "user not found",
		},
		{
			name: "ErrDomainNotFound",
			err:  happydns.ErrDomainNotFound,
			msg:  "domain not found",
		},
		{
			name: "ErrDomainLogNotFound",
			err:  happydns.ErrDomainLogNotFound,
			msg:  "domain log not found",
		},
		{
			name: "ErrProviderNotFound",
			err:  happydns.ErrProviderNotFound,
			msg:  "provider not found",
		},
		{
			name: "ErrSessionNotFound",
			err:  happydns.ErrSessionNotFound,
			msg:  "session not found",
		},
		{
			name: "ErrUserNotFound",
			err:  happydns.ErrUserNotFound,
			msg:  "user not found",
		},
		{
			name: "ErrUserAlreadyExist",
			err:  happydns.ErrUserAlreadyExist,
			msg:  "user already exists",
		},
		{
			name: "ErrZoneNotFound",
			err:  happydns.ErrZoneNotFound,
			msg:  "zone not found",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.err.Error() != tt.msg {
				t.Errorf("Error message = %q; want %q", tt.err.Error(), tt.msg)
			}
		})
	}
}

File: web/src/lib/errors.test.ts
// This file is part of the happyDomain (R) project.
// Copyright (c) 2022-2026 happyDomain
//
// This program is offered under a commercial and under the AGPL license.
// For commercial licensing, contact us at <contact@happydomain.org>.
//
// For AGPL licensing:
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import { describe, it, expect, vi, beforeEach } from "vitest";

vi.mock("$lib/stores/usersession", () => ({
    refreshUserSession: vi.fn(),
}));

import { refreshUserSession } from "$lib/stores/usersession";
import {
    NotAuthorizedError,
    ProviderNoDomainListingSupport,
    handleEmptyApiResponse,
    handleAuthApiResponse,
    handleApiResponse,
} from "./errors";

const mockedRefreshUserSession = vi.mocked(refreshUserSession);

beforeEach(() => {
    mockedRefreshUserSession.mockReset();
});

describe("NotAuthorizedError", () => {
    it("is an instance of Error", () => {
        const err = new NotAuthorizedError("nope");
        expect(err).toBeInstanceOf(Error);
    });

    it("sets the message and name", () => {
        const err = new NotAuthorizedError("nope");
        expect(err.message).toBe("nope");
        expect(err.name).toBe("NotAuthorizedError");
    });
});

describe("ProviderNoDomainListingSupport", () => {
    it("is an instance of Error", () => {
        const err = new ProviderNoDomainListingSupport("no listing");
        expect(err).toBeInstanceOf(Error);
    });

    it("sets the message and name", () => {
        const err = new ProviderNoDomainListingSupport("no listing");
        expect(err.message).toBe("no listing");
        expect(err.name).toBe("ProviderNoDomainListingSupport");
    });
});

describe("handleEmptyApiResponse", () => {
    it("returns true for a 204 response without reading the body", async () => {
        const res = new Response(null, { status: 204 });
        const jsonSpy = vi.spyOn(res, "json");

        await expect(handleEmptyApiResponse(res)).resolves.toBe(true);
        expect(jsonSpy).not.toHaveBeenCalled();
    });

    it("delegates to handleApiResponse for non-204 responses", async () => {
        const res = new Response(JSON.stringify(true), { status: 200 });

        await expect(handleEmptyApiResponse(res)).resolves.toBe(true);
    });

    it("delegates a 401 to the session refresh flow", async () => {
        mockedRefreshUserSession.mockRejectedValue(new Error("session expired"));
        const res = new Response(null, { status: 401 });

        await expect(handleEmptyApiResponse(res)).rejects.toThrow(NotAuthorizedError);
        expect(mockedRefreshUserSession).toHaveBeenCalled();
    });
});

describe("handleAuthApiResponse", () => {
    it("returns the parsed JSON body when the response is ok", async () => {
        const res = new Response(JSON.stringify({ foo: "bar" }), { status: 200 });

        await expect(handleAuthApiResponse<{ foo: string }>(res)).resolves.toEqual({
            foo: "bar",
        });
    });

    it("throws ProviderNoDomainListingSupport for the matching errmsg", async () => {
        const res = new Response(
            JSON.stringify({ errmsg: "the provider doesn't support domain listing" }),
            { status: 400 },
        );

        await expect(handleAuthApiResponse(res)).rejects.toThrow(ProviderNoDomainListingSupport);
    });

    it("throws a generic Error with errmsg for other error messages", async () => {
        const res = new Response(JSON.stringify({ errmsg: "something went wrong" }), {
            status: 400,
        });

        await expect(handleAuthApiResponse(res)).rejects.toThrow("something went wrong");
    });

    it("does not throw ProviderNoDomainListingSupport for an unrelated errmsg", async () => {
        const res = new Response(JSON.stringify({ errmsg: "something went wrong" }), {
            status: 400,
        });

        try {
            await handleAuthApiResponse(res);
            throw new Error("expected handleAuthApiResponse to reject");
        } catch (err) {
            expect(err).not.toBeInstanceOf(ProviderNoDomainListingSupport);
        }
    });

    it("throws a status-based Error when the body has no errmsg", async () => {
        const res = new Response(JSON.stringify({}), { status: 503, statusText: "" });

        await expect(handleAuthApiResponse(res)).rejects.toThrow("A 503 error occurs.");
    });
});

describe("handleApiResponse", () => {
    it("returns the parsed JSON body when the response is ok", async () => {
        const res = new Response(JSON.stringify({ foo: "bar" }), { status: 200 });

        await expect(handleApiResponse<{ foo: string }>(res)).resolves.toEqual({ foo: "bar" });
    });

    it("delegates to handleAuthApiResponse for non-401 error responses", async () => {
        const res = new Response(JSON.stringify({ errmsg: "boom" }), { status: 500 });

        await expect(handleApiResponse(res)).rejects.toThrow("boom");
        expect(mockedRefreshUserSession).not.toHaveBeenCalled();
    });

    it("refreshes the session on a 401 and returns the parsed body on success", async () => {
        mockedRefreshUserSession.mockResolvedValue(undefined as never);
        const res = new Response(JSON.stringify({ foo: "bar" }), { status: 401 });

        await expect(handleApiResponse<{ foo: string }>(res)).resolves.toEqual({ foo: "bar" });
        expect(mockedRefreshUserSession).toHaveBeenCalled();
    });

    it("wraps a refresh failure that is an Error in NotAuthorizedError", async () => {
        mockedRefreshUserSession.mockRejectedValue(new Error("refresh failed"));
        const res = new Response(null, { status: 401 });

        await expect(handleApiResponse(res)).rejects.toThrow(NotAuthorizedError);
        await expect(handleApiResponse(res)).rejects.toThrow("refresh failed");
    });

    it("rethrows a non-Error thrown by the session refresh as-is", async () => {
        mockedRefreshUserSession.mockRejectedValue("plain string failure");
        const res = new Response(null, { status: 401 });

        await expect(handleApiResponse(res)).rejects.toBe("plain string failure");
    });
});
